File: worker/src/pulse.py
"""Global Anxiety Pulse - visitor tracking and analytics.

Tracks visitor activity by country to show:
- How many people are watching (last 10 minutes)
- Activity multiplier vs baseline
- Country-specific surge detection
"""

import json
import logging
from datetime import datetime, timezone

log = logging.getLogger("aegis.pulse")

# Pulse data stored in separate R2 key
PULSE_KEY = "pulse.json"

# How long to keep visits (10 minutes in ms)
VISIT_WINDOW_MS = 10 * 60 * 1000

# Country code to flag emoji mapping
COUNTRY_FLAGS = {
    "IL": "🇮🇱", "US": "🇺🇸", "DE": "🇩🇪", "GB": "🇬🇧", "IR": "🇮🇷",
    "FR": "🇫🇷", "NL": "🇳🇱", "AE": "🇦🇪", "SA": "🇸🇦", "JO": "🇯🇴",
    "EG": "🇪🇬", "TR": "🇹🇷", "IN": "🇮🇳", "PK": "🇵🇰", "CA": "🇨🇦",
    "AU": "🇦🇺", "BR": "🇧🇷", "RU": "🇷🇺", "CN": "🇨🇳", "JP": "🇯🇵",
    "KR": "🇰🇷", "IT": "🇮🇹", "ES": "🇪🇸", "PL": "🇵🇱", "UA": "🇺🇦",
    "SE": "🇸🇪", "NO": "🇳🇴", "FI": "🇫🇮", "DK": "🇩🇰", "CH": "🇨🇭",
    "AT": "🇦🇹", "BE": "🇧🇪", "GR": "🇬🇷", "PT": "🇵🇹", "CZ": "🇨🇿",
    "RO": "🇷🇴", "HU": "🇭🇺", "IE": "🇮🇪", "SG": "🇸🇬", "MY": "🇲🇾",
    "TH": "🇹🇭", "VN": "🇻🇳", "PH": "🇵🇭", "ID": "🇮🇩", "MX": "🇲🇽",
    "AR": "🇦🇷", "CL": "🇨🇱", "CO": "🇨🇴", "ZA": "🇿🇦", "NG": "🇳🇬",
    "KE": "🇰🇪", "IQ": "🇮🇶", "SY": "🇸🇾", "LB": "🇱🇧", "KW": "🇰🇼",
    "QA": "🇶🇦", "BH": "🇧🇭", "OM": "🇴🇲", "YE": "🇾🇪", "AF": "🇦🇫",
    "XX": "🌍",  # Unknown
}

# Default baselines (estimated hourly visitors by country)
DEFAULT_BASELINES = {
    "total_hourly": 100,  # Default expected visitors per 10 min
    "by_country": {
        "US": 35, "IL": 15, "DE": 10, "GB": 10, "IR": 8,
        "FR": 5, "NL": 4, "CA": 4, "AU": 3, "IN": 3,
    }
}


def get_flag(country_code: str) -> str:
    """Get flag emoji for country code."""
    return COUNTRY_FLAGS.get(country_code, "🌍")


async def load_pulse_data(env) -> dict:
    """Load pulse data from R2, return empty structure if not found."""
    try:
        obj = await env.DATA_BUCKET.get(PULSE_KEY)
        if obj is not None:
            text = await obj.text()
            return json.loads(text)
    except Exception as e:
        log.warning("Failed to load pulse data: %s", e)

    # Return empty structure
    return {
        "visits": [],
        "baselines": DEFAULT_BASELINES,
    }


async def save_pulse_data(env, pulse_data: dict):
    """Save pulse data to R2."""
    try:
        payload = json.dumps(pulse_data)
        await env.DATA_BUCKET.put(PULSE_KEY, payload)
    except Exception as e:
        log.error("Failed to save pulse data: %s", e)


def trim_old_visits(visits: list, now_ms: int) -> list:
    """Remove visits older than VISIT_WINDOW_MS."""
    cutoff = now_ms - VISIT_WINDOW_MS
    return [v for v in visits if v.get("ts", 0) > cutoff]


async def log_visit(env, country_code: str) -> dict:
    """Log a visit and return current pulse stats.

    Args:
        env: Cloudflare Worker environment with R2 binding
        country_code: 2-letter country code from cf-ipcountry header

    Returns:
        Calculated pulse stats dict
    """
    now_ms = int(datetime.now(timezone.utc).timestamp() * 1000)

    # Load existing pulse data
    pulse_data = await load_pulse_data(env)

    # Trim old visits
    visits = trim_old_visits(pulse_data.get("visits", []), now_ms)

    # Add this visit
    visits.append({"ts": now_ms, "cc": country_code or "XX"})

    # Keep only last 15 minutes max (buffer beyond 10 min window)
    max_visits = 10000  # Safety limit
    if len(visits) > max_visits:
        visits = visits[-max_visits:]

    # Save updated visits
    pulse_data["visits"] = visits
    await save_pulse_data(env, pulse_data)

    # Calculate and return stats
    return calculate_pulse_stats(visits, pulse_data.get("baselines", DEFAULT_BASELINES), now_ms)


def calculate_pulse_stats(visits: list, baselines: dict, now_ms: int) -> dict:
    """Calculate pulse statistics from visits.

    Args:
        visits: List of visit records [{ts, cc}, ...]
        baselines: Baseline expectations
        now_ms: Current timestamp in milliseconds

    Returns:
        Pulse stats dict for frontend
    """
    # Filter to last 10 minutes
    cutoff = now_ms - VISIT_WINDOW_MS
    recent_visits = [v for v in visits if v.get("ts", 0) > cutoff]

    # Count total watching
    watching_now = len(recent_visits)

    # Count by country
    country_counts = {}
    for v in recent_visits:
        cc = v.get("cc", "XX")
        country_counts[cc] = country_counts.get(cc, 0) + 1

    # Calculate activity multiplier
    baseline_total = baselines.get("total_hourly", 100)
    if baseline_total > 0:
        activity_multiplier = round(watching_now / baseline_total, 1)
    else:
        activity_multiplier = 1.0

    # Determine activity level
    if activity_multiplier <= 1.2:
        activity_level = "normal"
    elif activity_multiplier <= 2.0:
        activity_level = "elevated"
    elif activity_multiplier <= 3.0:
        activity_level = "high"
    else:
        activity_level = "surging"

    # Calculate country surges
    country_baselines = baselines.get("by_country", {})
    countries = []

    for cc, count in sorted(country_counts.items(), key=lambda x: x[1], reverse=True):
        baseline = country_baselines.get(cc, 5)  # Default baseline of 5
        surge = round(count / baseline, 2) if baseline > 0 else 1.0

        countries.append({
            "cc": cc,
            "flag": get_flag(cc),
            "count": count,
            "surge": surge,
        })

    # Israel stats (always include even if 0)
    israel_data = next((c for c in countries if c["cc"] == "IL"), None)
    if israel_data:
        israel = {
            "count": israel_data["count"],
            "surge": israel_data["surge"],
        }
    else:
        israel = {"count": 0, "surge": 0}

    # Other countries (exclude Israel, filter to surge >= 1.5 or top 6)
    other_countries = [c for c in countries if c["cc"] != "IL"]

    # Show countries with surge >= 1.5, or top 6 if not enough
    surging_countries = [c for c in other_countries if c["surge"] >= 1.5]
    if len(surging_countries) < 4:
        surging_countries = other_countries[:6]
    else:
        surging_countries = surging_countries[:6]

    return {
        "watching_now": watching_now,
        "activity_multiplier": activity_multiplier,
        "activity_level": activity_level,
        "israel": israel,
        "countries": surging_countries,
        "total_countries": len(country_counts),
    }


async def get_pulse_stats_only(env) -> dict:
    """Get pulse stats without logging a new visit.

    Used by scheduled handler or when we just want to read stats.
    """
    now_ms = int(datetime.now(timezone.utc).timestamp() * 1000)
    pulse_data = await load_pulse_data(env)
    visits = pulse_data.get("visits", [])
    baselines = pulse_data.get("baselines", DEFAULT_BASELINES)

    return calculate_pulse_stats(visits, baselines, now_ms)

File: backend/internal/pulse/pulse.go
package pulse

import (
	"sync"
	"time"
)

// Visit represents a single page visit.
type Visit struct {
	Timestamp   time.Time
	CountryCode string
}

// CountryStats holds statistics for a single country.
type CountryStats struct {
	CC    string `json:"cc"`
	Flag  string `json:"flag"`
	Count int    `json:"count"`
	Surge float64 `json:"surge"`
}

// IsraelStats holds Israel-specific statistics.
type IsraelStats struct {
	Count int     `json:"count"`
	Surge float64 `json:"surge"`
}

// Stats is the pulse data returned to the frontend.
type Stats struct {
	WatchingNow        int            `json:"watching_now"`
	ActivityMultiplier float64        `json:"activity_multiplier"`
	ActivityLevel      string         `json:"activity_level"`
	Israel             IsraelStats    `json:"israel"`
	Countries          []CountryStats `json:"countries"`
	TotalCountries     int            `json:"total_countries"`
}

// Tracker tracks visitor activity with a sliding time window.
type Tracker struct {
	mu         sync.RWMutex
	visits     []Visit
	window     time.Duration
	maxVisits  int
	baselines  map[string]int
	baseTotal  int
}

// Country code to flag emoji mapping.
var countryFlags = map[string]string{
	"IL": "🇮🇱", "US": "🇺🇸", "DE": "🇩🇪", "GB": "🇬🇧", "IR": "🇮🇷",
	"FR": "🇫🇷", "NL": "🇳🇱", "AE": "🇦🇪", "SA": "🇸🇦", "JO": "🇯🇴",
	"EG": "🇪🇬", "TR": "🇹🇷", "IN": "🇮🇳", "PK": "🇵🇰", "CA": "🇨🇦",
	"AU": "🇦🇺", "BR": "🇧🇷", "RU": "🇷🇺", "CN": "🇨🇳", "JP": "🇯🇵",
	"KR": "🇰🇷", "IT": "🇮🇹", "ES": "🇪🇸", "PL": "🇵🇱", "UA": "🇺🇦",
	"SE": "🇸🇪", "NO": "🇳🇴", "FI": "🇫🇮", "DK": "🇩🇰", "CH": "🇨🇭",
	"AT": "🇦🇹", "BE": "🇧🇪", "GR": "🇬🇷", "PT": "🇵🇹", "CZ": "🇨🇿",
	"RO": "🇷🇴", "HU": "🇭🇺", "IE": "🇮🇪", "SG": "🇸🇬", "MY": "🇲🇾",
	"TH": "🇹🇭", "VN": "🇻🇳", "PH": "🇵🇭", "ID": "🇮🇩", "MX": "🇲🇽",
	"AR": "🇦🇷", "CL": "🇨🇱", "CO": "🇨🇴", "ZA": "🇿🇦", "NG": "🇳🇬",
	"KE": "🇰🇪", "IQ": "🇮🇶", "SY": "🇸🇾", "LB": "🇱🇧", "KW": "🇰🇼",
	"QA": "🇶🇦", "BH": "🇧🇭", "OM": "🇴🇲", "YE": "🇾🇪", "AF": "🇦🇫",
}

// Default baselines (expected visitors per 10-minute window).
var defaultBaselines = map[string]int{
	"US": 35, "IL": 15, "DE": 10, "GB": 10, "IR": 8,
	"FR": 5, "NL": 4, "CA": 4, "AU": 3, "IN": 3,
}

const defaultBaseTotal = 100

// NewTracker creates a new pulse tracker.
func NewTracker() *Tracker {
	return &Tracker{
		visits:    make([]Visit, 0, 1000),
		window:    10 * time.Minute,
		maxVisits: 10000,
		baselines: defaultBaselines,
		baseTotal: defaultBaseTotal,
	}
}

// getFlag returns the flag emoji for a country code.
func getFlag(cc string) string {
	if flag, ok := countryFlags[cc]; ok {
		return flag
	}
	return "🌍"
}

// LogVisit records a visit and returns current stats.
func (t *Tracker) LogVisit(countryCode string) Stats {
	now := time.Now()

	if countryCode == "" {
		countryCode = "XX"
	}

	t.mu.Lock()
	// Trim old visits
	t.trimOldVisits(now)

	// Add new visit
	t.visits = append(t.visits, Visit{
		Timestamp:   now,
		CountryCode: countryCode,
	})

	// Enforce max visits limit
	if len(t.visits) > t.maxVisits {
		t.visits = t.visits[len(t.visits)-t.maxVisits:]
	}

	// Calculate stats while holding lock
	stats := t.calculateStats(now)
	t.mu.Unlock()

	return stats
}

// GetStats returns current stats without logging a visit.
func (t *Tracker) GetStats() Stats {
	now := time.Now()

	t.mu.Lock()
	t.trimOldVisits(now)
	stats := t.calculateStats(now)
	t.mu.Unlock()

	return stats
}

// trimOldVisits removes visits outside the time window.
// Must be called with lock held.
func (t *Tracker) trimOldVisits(now time.Time) {
	cutoff := now.Add(-t.window)
	idx := 0
	for i, v := range t.visits {
		if v.Timestamp.After(cutoff) {
			idx = i
			break
		}
		if i == len(t.visits)-1 {
			idx = len(t.visits)
		}
	}
	if idx > 0 {
		t.visits = t.visits[idx:]
	}
}

// calculateStats computes pulse statistics from current visits.
// Must be called with lock held.
func (t *Tracker) calculateStats(now time.Time) Stats {
	cutoff := now.Add(-t.window)

	// Count visits by country
	countryCounts := make(map[string]int)
	for _, v := range t.visits {
		if v.Timestamp.After(cutoff) {
			countryCounts[v.CountryCode]++
		}
	}

	watchingNow := 0
	for _, count := range countryCounts {
		watchingNow += count
	}

	// Calculate activity multiplier
	var activityMultiplier float64
	if t.baseTotal > 0 {
		activityMultiplier = float64(watchingNow) / float64(t.baseTotal)
		// Round to 1 decimal
		activityMultiplier = float64(int(activityMultiplier*10)) / 10
	} else {
		activityMultiplier = 1.0
	}

	// Determine activity level
	var activityLevel string
	switch {
	case activityMultiplier <= 1.2:
		activityLevel = "normal"
	case activityMultiplier <= 2.0:
		activityLevel = "elevated"
	case activityMultiplier <= 3.0:
		activityLevel = "high"
	default:
		activityLevel = "surging"
	}

	// Calculate country stats with surge
	type countryData struct {
		cc    string
		count int
		surge float64
	}
	var countries []countryData

	for cc, count := range countryCounts {
		baseline := t.baselines[cc]
		if baseline == 0 {
			baseline = 5 // Default baseline
		}
		surge := float64(count) / float64(baseline)
		// Round to 2 decimals
		surge = float64(int(surge*100)) / 100

		countries = append(countries, countryData{
			cc:    cc,
			count: count,
			surge: surge,
		})
	}

	// Sort by count descending
	for i := 0; i < len(countries); i++ {
		for j := i + 1; j < len(countries); j++ {
			if countries[j].count > countries[i].count {
				countries[i], countries[j] = countries[j], countries[i]
			}
		}
	}

	// Israel stats (always include)
	israel := IsraelStats{Count: 0, Surge: 0}
	for _, c := range countries {
		if c.cc == "IL" {
			israel.Count = c.count
			israel.Surge = c.surge
			break
		}
	}

	// Other countries (exclude Israel)
	var otherCountries []countryData
	for _, c := range countries {
		if c.cc != "IL" {
			otherCountries = append(otherCountries, c)
		}
	}

	// Filter to surging countries or top 6
	var displayCountries []CountryStats
	surgingCount := 0
	for _, c := range otherCountries {
		if c.surge >= 1.5 {
			surgingCount++
		}
	}

	if surgingCount >= 4 {
		// Show surging countries (up to 6)
		for _, c := range otherCountries {
			if c.surge >= 1.5 && len(displayCountries) < 6 {
				displayCountries = append(displayCountries, CountryStats{
					CC:    c.cc,
					Flag:  getFlag(c.cc),
					Count: c.count,
					Surge: c.surge,
				})
			}
		}
	} else {
		// Show top 6 by count
		for i, c := range otherCountries {
			if i >= 6 {
				break
			}
			displayCountries = append(displayCountries, CountryStats{
				CC:    c.cc,
				Flag:  getFlag(c.cc),
				Count: c.count,
				Surge: c.surge,
			})
		}
	}

	return Stats{
		WatchingNow:        watchingNow,
		ActivityMultiplier: activityMultiplier,
		ActivityLevel:      activityLevel,
		Israel:             israel,
		Countries:          displayCountries,
		TotalCountries:     len(countryCounts),
	}
}
